### ISSUE TITLE
Explicit literal types in serialized RDF

### ISSUE THREAD

**Issue by rmarkelo** [SynBioDex/pySBOL3#113]
Currently, when a document is serialized as RDF, literal types for datatype property objects are lost. For example, the value 4 for an integer datatype property is written as 4 rather than "4"^^xsd:int. When the file is read in using rdflib, it looks like all the literals are interpreted as strings. This causes problems for SHACL validation and could potentially cause other problems.

**Comment by tcmitchell** [comment]
Is it possible this has something to do with the turtle format, which I think you're using? Can you try to format the file in ntriples (`sbol3.NTRIPLES`) and see if the output shows your property with the integer type?

Here is an example using RDFlib, independent of pySBOL3, which demonstrates that RDFlib does not include the type in turtle and does include the type in ntriples.

```python
$ python3
Python 3.8.5 (v3.8.5:580fbb018f, Jul 20 2020, 12:11:27) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import rdflib
RDFLib Version: 5.0.0
>>> bob = rdflib.URIRef("http://example.org/people/Bob")
>>> g = rdflib.Graph()
>>> g.add((bob, rdflib.namespace.FOAF.age, rdflib.Literal(42)))
>>> print(g.serialize(format="turtle").decode("utf-8"))
@prefix ns1: <http://xmlns.com/foaf/0.1/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<http://example.org/people/Bob> ns1:age 42 .


>>> print(g.serialize(format="nt").decode("utf-8"))
<http://example.org/people/Bob> <http://xmlns.com/foaf/0.1/age> "42"^^<http://www.w3.org/2001/XMLSchema#integer> .


>>>
```

It may also be possible to ask RDFlib to include the type annotations in turtle. I don't know enough about the turtle format to know one way or the other.

**Comment by tcmitchell** [comment]
I don't see a way to include the datatype annotation you desire in the RDFlib turtle output. I'd be happy to add that capability and/or expose an ability to add datatype annotations if you can show me how to manipulate the [RDFlib turtle serializer](https://github.com/RDFLib/rdflib/blob/master/rdflib/plugins/serializers/turtle.py) to do so. I don't see that capability via manual code inspection and the documentation is limited.

My read of the [turtle specification](https://www.w3.org/TeamSubmission/turtle/), Section 2.4, indicates that the datatype annotation is optional for integers, so I think RDFlib is writing valid turtle. I'm not sure why your validator is requiring the datatype annotation.

**Comment by bbartley** [comment]
Here is an example of turtle serialization with explicit integer type.  Interestingly, I can't seem to get it to serialize an explicit string type.  However, per W3C specification of Turtle, writing just "my_string" implicitly means it's xsd:string
```
import rdflib
  
my_foo = rdflib.URIRef('http://example.org#my_foo')
has_bar = rdflib.URIRef('http://example.org#has_bar')
bar = rdflib.Literal(4, datatype='http://www.w3.org/TR/xmlschema11-2/#int')
g = rdflib.Graph()
g.add((my_foo, has_bar, bar))
g.serialize('foo.ttl', format='ttl')
```
This results in the following serialization:
```
@prefix ns1: <http://example.org#> .

ns1:my_foo ns1:has_bar "4"^^<http://www.w3.org/TR/xmlschema11-2/#int> .
```

**Comment by bbartley** [comment]
Actually, I take that back.  I _can_ explicitly indicate string type.  Not sure why it didn't seem to work previously:
```
import rdflib
  
my_foo = rdflib.URIRef('http://example.org#my_foo')
has_bar = rdflib.URIRef('http://example.org#has_bar')
bar = rdflib.Literal(4, datatype='http://www.w3.org/TR/xmlschema11-2/#string')
g = rdflib.Graph()
g.add((my_foo, has_bar, bar))
g.serialize('foo.ttl', format='ttl')
```
Serialization:
```
@prefix ns1: <http://example.org#> .

ns1:my_foo ns1:has_bar "4"^^<http://www.w3.org/TR/xmlschema11-2/#string> .
```

**Comment by tcmitchell** [comment]
I guess this gets at what the correct expansion of `xsd:` is. RDFlib thinks the XMLSchema Namespace is `http://www.w3.org/2001/XMLSchema#`.

```python
>>> import rdflib
RDFLib Version: 5.0.0
>>> x = rdflib.Literal(42)
>>> x.datatype
rdflib.term.URIRef('http://www.w3.org/2001/XMLSchema#integer')
>>> rdflib.namespace.XSD
Namespace('http://www.w3.org/2001/XMLSchema#')
>>> 
```

You're using a different namespace for the `xsd:` expansion, `http://www.w3.org/TR/xmlschema11-2/#`. In effect, you're labeling the value with a different data type. Does `http://www.w3.org/TR/xmlschema11-2/#int` work with the pySHACL validation?

Do we want to have pySBOL3 bypass the RDFlib concept of XSD being `http://www.w3.org/2001/XMLSchema#` and replace it with `http://www.w3.org/TR/xmlschema11-2/#`? That seems risky, and I don't know how portable it will be.

-----
But wait, there's more.

* The [SBOL 3.0 specification](https://sbolstandard.org/wp-content/uploads/2020/04/SBOL3.0specification.pdf), Section 5.2, page 12, specifically says that the datatype for integers is `http://www.w3.org/TR/xmlschema11-2/#integer`. That [has been changed in the draft 3.1 specification](https://github.com/SynBioDex/SBOL-specification/issues/395) to `http://www.w3.org/2001/XMLSchema#integer`. 
* The [turtle specification](https://www.w3.org/TeamSubmission/turtle/), in Section 2.4. Abbreviating common datatypes, expands `xsd:integer` to `http://www.w3.org/2001/XMLSchema#integer`.

The SBOL specification issue concerning the proper namespace links to the appropriate places in the XML Schema documentation that indicates that the proper namespace should be `http://www.w3.org/2001/XMLSchema#`, not `http://www.w3.org/TR/xmlschema11-2/#` as you have specified above:
* https://www.w3.org/TR/xmlschema11-2/#namespaces
* https://www.w3.org/TR/xmlschema11-2/#sec-prim-nxsd

The bottom line is that I think we should stick with the standard namespace (`http://www.w3.org/2001/XMLSchema#`) and thus I think we're back where we started. I don't know how to get RDFlib to add the datatype annotation `http://www.w3.org/2001/XMLSchema#integer` to the turtle output.

**Comment by rmarkelo** [comment]
Since the explicit literal types appear with NTriples serialization, but not with Turtle, this must be a quirk of rdflib. If we need to do SHACL validation of graphs that are read in from files, using NTriples serialization is a reasonable workaround.

As far as I know http://www.w3.org/2001/XMLSchema# has always been the standard  XMLSchema namespace. We should use that and spurn any other namespace.
